### DISCUSSION TITLE
Exception handling

### DISCUSSION THREAD

**Discussion by P72B** [davidepianca98/KMQTT#57] [Q&A]
Hi,
I have hard times with the exception handling for the kmqtt library. What I'm currently doing is:
`CoroutineScope(Dispatchers.IO + exceptionHandler).launch {
    mqttClient.step()
}`
The problem: I don't get the correct exception in the exceptionHandler. I see in my coroutine handler an IllegalArgumentException but I see a MQTTException with BAD_USER_NAME_OR_PASSWORD in the private field `lastException` of the mqttClient.

<img width="1008" alt="Screenshot 2024-08-01 104204" src="https://github.com/user-attachments/assets/e2032779-4165-4f42-83da-16ddbbd96b70">

How can I access/handle the MQTTException correctly? Can we make the private field `lastException` public? Or can we pass a new parameter in the constructor of MQTTClient to handle exceptions?

Thank you for the help in advance

**Comment by davidepianca98** [comment]
Hello, I prepared this example to test and it looks like I am getting the exception in handler:

```
public fun main() {
    val client = MQTTClient(
        MQTTVersion.MQTT5,
        "test.mosquitto.org",
        1884,
        null,
        userName = "aaa",
        password = "bbb".encodeToByteArray().toUByteArray(),
        keepAlive = 15
    ) {
        println(it.payload?.toByteArray()?.decodeToString())
    }

    runBlocking {
        val handler = CoroutineExceptionHandler { _, exception ->
            println("CoroutineExceptionHandler got $exception")
        }

        val job = CoroutineScope(Dispatchers.IO + handler).launch {
            while (true) {
                client.step()
                delay(100)
            }
        }
        job.join()
    }
}
```

The output is `CoroutineExceptionHandler got mqtt.MQTTException: NOT_AUTHORIZED`
